### PR TITLE
Default device_map to None on MPS in create_model_from_path

### LINKED ISSUE #7220: Apple Silicon (MPS): create_model_from_path can segfault while loading bf16 checkpoints
### Reproduction

On Apple Silicon, trainers loading a bf16 checkpoint from a model ID with the default initialization arguments can segfault while loading weights.

```python
from datasets import Dataset
from trl import GRPOConfig, GRPOTrainer

trainer = GRPOTrainer(
    model="trl-internal-testing/tiny-Qwen3ForCausalLM",
    reward_funcs=lambda completions, **kwargs: [0.0] * len(completions),
    args=GRPOConfig(output_dir="out", per_device_train_batch_size=2, num_generations=2, report_to="none"),
    train_dataset=Dataset.from_dict({"prompt": ["hi", "hello"]}),
)
```

```console
$ python -X faulthandler repro.py
Fatal Python error: Segmentation fault
```

This exited with code 139 in 4 of 4 fresh processes. The same happens with `Qwen/Qwen2.5-0.5B-Instruct`. It reduces to TRL's loader alone (3 of 3 runs):

```python
from trl.trainer.utils import create_model_from_path

create_model_from_path("trl-internal-testing/tiny-Qwen3ForCausalLM")
```

#### Cause

[`create_model_from_path`](https://github.com/huggingface/trl/blob/v1.13.0/trl/trainer/utils.py#L1188) defaults `device_map` to `"auto"` unless the device is CPU, and [defaults `dtype` to `float32`](https://github.com/huggingface/trl/blob/v1.13.0/trl/trainer/utils.py#L1175):

```python
dtype = kwargs.get("dtype", "float32")
...
kwargs["device_map"] = None if PartialState().device.type == "cpu" else "auto"
```

On a Mac, `PartialState().device.type` is `mps`. A bf16 checkpoint loaded with these defaults therefore meets the conditions in huggingface/transformers#48029: the async loader can crash while converting bf16 weights to `float32` and loading them onto MPS.

Not every load crashes. In my tests, `device_map=None`, disabling async loading, or matching `dtype` to the checkpoint each avoided it. Based on #48029, a float32 checkpoint should not trigger it either. `create_model_from_path` is used by SFT, DPO, KTO, RLOO, GRPO, Reward, Distillation, and several experimental trainers, so the default is not specific to GRPO.

#### Workarounds

All three of these allowed `GRPOTrainer` to construct with the 0.5B model:

- Set `model_init_kwargs={"device_map": None}` in the trainer config.
- Set `model_init_kwargs={"dtype": "auto"}` where MPS supports the checkpoint's dtype. I verified loading and trainer construction, not training with this setting.
- Set `HF_DEACTIVATE_ASYNC_LOAD=1` before starting Python.

Preloading the policy model only partly avoids the problem. With full fine-tuning, no PEFT, and `beta > 0`, GRPO still [creates its reference model through `create_model_from_path`](https://github.com/huggingface/trl/blob/v1.13.0/trl/trainer/grpo_trainer.py#L988) and crashes. `model_init_kwargs={"device_map": None}` also avoids this reference-model crash, although GRPO first [warns that those kwargs will be ignored](https://github.com/huggingface/trl/blob/v1.13.0/trl/trainer/grpo_trainer.py#L349) because the policy is preloaded; the reference-model path subsequently uses them.

#### Faulthandler output

The process exits via SIGSEGV, so there is no ordinary Python exception traceback. In 3 of 4 runs, faulthandler produced multiple interleaved `Fatal Python error` headers, consistent with loader workers faulting concurrently. The readable stacks were:

```
Current thread (most recent call first):
  File "<site-packages>/transformers/core_model_loading.py", line 1240 in _materialize_copy
  File "<site-packages>/transformers/core_model_loading.py", line 1262 in _job
  File "<python3.12>/concurrent/futures/thread.py", line 59 in run
  File "<python3.12>/concurrent/futures/thread.py", line 93 in _worker

Main thread (most recent call first):
  File "<python3.12>/concurrent/futures/_base.py", line 451 in result
  File "<site-packages>/transformers/core_model_loading.py", line 970 in materialize_tensors
  File "<site-packages>/transformers/core_model_loading.py", line 1008 in convert
  File "<site-packages>/transformers/core_model_loading.py", line 1754 in convert_and_load_state_dict_in_model
  File "<site-packages>/transformers/modeling_utils.py", line 4456 in _load_pretrained_model
  File "<site-packages>/transformers/modeling_utils.py", line 4313 in from_pretrained
  File "<site-packages>/trl/trainer/utils.py", line 1203 in create_model_from_path
  File "<site-packages>/trl/trainer/grpo_trainer.py", line 343 in __init__
  File "repro.py", line 4 in <module>
```

#### Possible fix

For ordinary single-device MPS training, defaulting to `device_map=None` avoids the crashing Transformers path and leaves placement to Accelerate or the Trainer:

```diff
     if "device_map" not in kwargs:
-        kwargs["device_map"] = None if PartialState().device.type == "cpu" else "auto"
+        # device_map="auto" on MPS can crash Transformers' async loader (huggingface/transformers#48029)
+        kwargs["device_map"] = None if PartialState().device.type in ("cpu", "mps") else "auto"
```

The trade-off is that MPS models would load on CPU first, and automatic CPU or disk offload for models too large for accelerator memory would be lost. #4509 introduced the `"auto"` default to avoid the RAM-to-VRAM move, based on CUDA measurements. I have not measured the load-time difference on MPS. If huggingface/transformers#48029 is fixed upstream, this guard could be removed.

I can open a PR with this guard if that approach is acceptable.

#### What I tested

TRL 1.13.0 and Transformers 5.17.0; each case ran in a fresh process.

| Case | Result |
| --- | --- |
| Tiny checkpoint: loader / GRPO | Segfault, 3/3 / 4/4 |
| GRPO with 0.5B model and defaults | Segfault |
| Same with `device_map=None` | Constructs |
| Same with `dtype="auto"` | Constructs |
| Same with `HF_DEACTIVATE_ASYNC_LOAD=1` | Constructs |
| Preloaded policy, `beta=0.04`, no PEFT: default / `device_map=None` | Segfault / constructs; reference model on `mps:0` |
| Proposed patch: model ID / preloaded policy with reference model | Both construct |

These tests cover loading and trainer construction. I did not run training steps with the patch or test other trainers.

#### Related

- huggingface/transformers#48029: underlying loader crash
- #4509: introduced the `device_map="auto"` default
- #6295: added the CPU exception to that default

### System Info

```text
- Platform: macOS-26.6.2-arm64-arm-64bit
- Python version: 3.12.12
- TRL version: 1.13.0
- PyTorch version: 2.14.0
- accelerator(s): MPS
- Transformers version: 5.17.0
- Accelerate version: 1.15.0
- Accelerate config: not found
- Datasets version: 5.0.1
- HF Hub version: 1.31.0
- bitsandbytes version: not installed
- DeepSpeed version: not installed
- Liger-Kernel version: not installed
- PEFT version: not installed
- vLLM version: not installed
```

Additional information:

- Hardware: Apple M4, 32 GB unified memory
- safetensors version: 0.8.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by dasashreeya**
# What does this PR do?

On Apple Silicon, `PartialState().device.type` is `"mps"`, so every trainer that loads a model through `create_model_from_path` (SFT, DPO, KTO, RLOO, GRPO, Reward, Distillation, and the GRPO reference model) used `device_map="auto"` together with the `float32` default. transformers' threaded loader then converts bf16 checkpoints on the MPS device, which segfaults or hangs (huggingface/transformers#48029, still open; the fix proposed there was closed).

This PR treats `"mps"` like `"cpu"` in the `device_map` default, so the model is loaded on CPU and moved to MPS by the trainer afterwards, exactly as on CPU today. An explicit `device_map` in `model_init_kwargs` is unchanged, and CUDA/XPU keep `"auto"`. It is the same one-line pattern as #6295, which added the CPU rule to this line.

Verified on an M4 Pro (macOS, Python 3.12, torch 2.14.0, transformers 5.17.0, safetensors 0.8.0), using the reproduction from the issue plus one training step:

| | `main` | this PR |
|---|---|---|
| Issue's `GRPOTrainer` snippet, `max_steps=1` | exit 139, `Fatal Python error` at `Loading weights 0/25`, loader received `device_map="auto"` | exit 0, loader received `device_map=None`, model on `mps:0` during and after training |
| Bare `create_model_from_path("trl-internal-testing/tiny-Qwen3ForCausalLM")` | hung for 16 minutes until the OS killed it for memory | loads in seconds |
| `tests/test_utils.py` | 157 passed | 161 passed (4 added) |

New tests in `tests/test_utils.py` (`TestCreateModelFromPath`) patch `PartialState` and `from_pretrained` and assert the `device_map` default for `cpu`, `mps`, `cuda`, and that an explicit `device_map` is kept. The `mps` case fails on `main`.

Not covered on purpose: `experimental/orpo`, `experimental/cpo` and `experimental/online_dpo` load through their own `AutoModelForCausalLM.from_pretrained` with a `device_map="auto"` default and never call `create_model_from_path`, so they keep the MPS hazard (as they did after #6295 for CPU). Routing them through the shared loader would be a separate change.

Fixes #7220

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. https://github.com/huggingface/trl/issues/7220
- [x] Did you make sure to update the documentation with your changes? (No documentation describes this default; the inline comment is updated.)
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

The diff and tests were produced with Claude Code under my direction. I reproduced the crash and the fix on my own Mac with the commands above, I have reviewed the change, and I will respond to review myself.

## Who can review?

@qgallouedec @sergiopaniego

### CODE REVIEW

**Comment by qgallouedec** [comment]
Closing: we don't review fully AI-generated PRs from first-time contributors, see the [AI usage policy](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#ai-usage-policy).
